### PR TITLE
Storage service

### LINKED ISSUE #22: File Storage Service Implementation
# This is a Feature Proposal

## Description

Today, every piece of software that wants to write to a file has a direct dependency on `fs` through `FileUtils`.
Eventually, this should be replaced, so that individual components don't have to know how to interact with the file system.

First though, we should implement a `FileStorageService` to abstract away this dependency.

Similar or dependent issues:

- #11

### PR DESCRIPTION
**Pull request by zmmille2**
## What did you implement:

Closes #22.
Implementation of a `FileStorageService`.
It can read files, overwrite files, and find a file from a starting directory recursively.

Turns the Interpolator into a service.
`ObjectService` template pattern, backed by a `StorageService` or a `StorageService` and an `InterpolatorService`.

## How did you implement it:

`FileStorageService` is backed by `fs`.
Other services with a dependency on `FileUtils` will be refactored to depend on `FileStorageService` in a future PR.

## Todos:

_**Note: Run `npm run check` to run all validation checks on proposed changes**_

- [X] Ensure there are no lint errors
       **Validate via `npm run lint`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [X] Write tests and confirm existing functionality is not broken
       **Validate via `npm test`**
- ~[ ] Write documentation~ This should not be exposed to the user.
- ~[ ] Pull request title and commit message follows [the conventional commit specification](https://www.conventionalcommits.org/en/v1.0.0-beta.2/#summary)~ I did not do this. I can update if need be.
- [X] Provide verification config / commands / resources (? I think I've done this. You can verify this works through the provided tests.
- [X] Requested reviews from at least 2 individuals
- [X] Assigned PR to myself
- [X] Added PR to correct project
- [X] This is ready for review

**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by zmmille2** [comment]
It's worth noting that the `Interpolator` in `utils` will also eventually get replaced.

File: src/models/object/objectService.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
+export abstract class ObjectService<T> {
```
Comment (tbarlow12) [inline src/models/object/objectService.ts:1]: Can we get some JSDocs for this class? just to provide context for what the goal of the `ObjectService` is and all of its public methods

File: src/models/storage/storageService.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export interface StorageService {
```
Comment (tbarlow12) [inline src/models/storage/storageService.ts:1]: nit: JSDocs - especially helpful for difference between `find` and `get`

File: src/services/object/storedInterpolatedObjectService.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { StoredObjectService } from ".";
+import { InterpolatorService, StorageService } from "../../models";
+
+export class StoredInterpolatedObjectService<T> extends StoredObjectService<T> {
+  private interpolatorService!: InterpolatorService;
```
Comment (tbarlow12) [inline src/services/object/storedInterpolatedObjectService.ts:5]: do you need the not null assertion here? seems like it's assigned in the constructor

File: src/services/object/storedObjectService.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { ObjectService, StorageService } from "../../models";
+
+export class StoredObjectService<T> extends ObjectService<T> {
```
Comment (tbarlow12) [inline src/services/object/storedObjectService.ts:3]: nit: JSDocs

**Review by tbarlow12 — COMMENTED** [review]
Looking good so far, I just want to have a little better understanding of what each of the services is intended to do. Some JSDocs would be helpful to gain that understanding and be able to review these with the intended context

File: src/services/object/storedInterpolatedObjectService.ts (hunk 3 above)
Comment (zmmille2) [reply src/services/object/storedInterpolatedObjectService.ts:5]: Yeah, this is a leftover from when I was messing around with function assignment, it should definitely come out

File: src/models/object/objectService.ts [hunk 5]
```diff
@@ -0,0 +1,126 @@
+/**
+ * A service following [the template pattern]{@link https://en.wikipedia.org/wiki/Template_method_pattern}
+ * for object storage and retrieval.
+ *
+ * This decouples storage and interpolation from object retrieval, so commands
+ * can delegate those responsibilities and focus on commanding.
```
Comment (tbarlow12) [inline src/models/object/objectService.ts:6]: 🤣

**Review by tbarlow12 — APPROVED** [review]
Beautiful. Maybe just add the `refactor: ` prefix to the squashed commit message

File: src/models/object/objectService.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+   *
+   * @param {string} name The name of the object to store.
+   * @param {T} obj The instance of the object to store.
+   * @param {string} location The place to store the object. If null, some
+   * default should be used.
+   *
+   * @returns {Promise<boolean>} Whether the set was successful.
+   * @template T
+   */
+  public async set(name: string, obj: T, location?: string): Promise<boolean> {
```
Comment (elizabethhalper) [inline src/models/object/objectService.ts:42]: I'm personally fine with this, but isn't it protocol for `set` methods to not return anything?

**Review by elizabethhalper — APPROVED** [review]
lgtm :) I second Tanner's comment about JSDocs, especially since "object" is very broad. I'm assuming things like workItems, or sprint/agile template stuff would be what we're reading as "objects" in the future so as long as the terminology and classes are intended to get more specific down the line I'm happy with this. Great work!!

**Pull request merged**